crates/forge/bin/cmd/test/mod.rs

```diff
@@ -537,151 +537,146 @@ async fn test(
-                                    (verbosity == 4 && result.status == TestStatus::Failure)
-                            }
-                            TraceKind::Execution => {
-                                verbosity > 3 ||
-                                    (verbosity == 3 && result.status == TestStatus::Failure)
-                            }
-                            _ => false,
-                        };
-
-                        // We decode the trace if we either need to build a gas report or we need
-                        // to print it
-                        if should_include || gas_reporting {
-                            decoder.decode(trace).await;
-                        }
-
-                        if should_include {
-                            decoded_traces.push(trace.to_string());
-                        }
-                    }
+            // Identify addresses in each trace
+            let mut builder = CallTraceDecoderBuilder::new()
+                .with_labels(result.labeled_addresses.iter().map(|(a, s)| (*a, s.clone())))
+                .with_events(local_identifier.events().cloned())
+                .with_verbosity(verbosity);
 
-                    if !decoded_traces.is_empty() {
-                        println!("Traces:");
-                        decoded_traces.into_iter().for_each(|trace| println!("{trace}"));
-                    }
+            // Signatures are of no value for gas reports
+            if !gas_reporting {
+                builder = builder.with_signature_identifier(sig_identifier.clone());
+            }
 
-                    if gas_reporting {
-                        gas_report.analyze(&result.traces);
+            let mut decoder = builder.build();
+
+            // Decode the traces
+            let mut decoded_traces = Vec::with_capacity(result.traces.len());
+            for (kind, trace) in &mut result.traces {
+                // verbosity:
+                // - 0..3: nothing
+                // - 3: only display traces for failed tests
+                // - 4: also display the setup trace for failed tests
+                // - 5..: display all traces for all tests
+                let should_include = match kind {
+                    TraceKind::Execution => {
+                        (verbosity == 3 && result.status.is_failure()) || verbosity >= 4
                     }
+                    TraceKind::Setup => {
+                        (verbosity == 4 && result.status.is_failure()) || verbosity >= 5
+                    }
+                    TraceKind::Deployment => false,
+                };
+
+                // We decode the trace if we either need to build a gas report or we need to print
+                // it
+                if should_include || gas_reporting {
+                    decoder.identify(trace, &mut local_identifier);
+                    decoder.identify(trace, &mut etherscan_identifier);
+                    decoder.decode(trace).await;
```

Comment:
This is the main change, moved from above the match to inside the if block

Reply:
this change is really weird, it speeds up significantly when there are a lot of traces, but not when there are little? I don't really know or have the resources to investigate, so I'll remove it from this PR